src/kernels/MIOpenDropoutHIP.cpp

```diff
@@ -0,0 +1,374 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+// Workaround to overcome redefinition errors while including rocrand header files directly
+#include "miopen_rocrand.hpp"
+
+#ifndef MIOPEN_USE_FP32
+#define MIOPEN_USE_FP32 0
+#endif
+
+#ifndef MIOPEN_USE_FP16
+#define MIOPEN_USE_FP16 0
+#endif
+
+#if MIOPEN_USE_FP16 == 1
+#define _FLOAT half
```

Comment:
Might be useful to replace these defines with kernels of different types. See the last lines in `src/kernels/gpu_reference_kernel/naive_conv.cpp`. 
The idea being that we instantiate the templated implementation with all supported data types and compile them all at once because they are in a single file.

Reply:
But if I don't need FP16 implementation at all? The whole compiled binary blob gets into the kernel cache for only the one specific kernel depicted by the network config. Network config says "FP32", so why we should spend time compiling "FP16" and waste space putting this FP16 stuff into the kernel cache.

Reply:
@CAHEK7 : My original thought was that compilation is expensive and if we compile all the variants of a kernel (in terms of data types etc) then we can amortize the cost somewhat. What you wrote means that we need to enhance the code object extraction so that we can pick specific kernel blobs out of the bigger binary blob and put those in the kernel cache.

Reply:
@amberhassaan, right now there is no way do distinguish two binary blobs except a network config. Then it tries to call some particular kernel (using its function name). If we want to use one blob for a multiple types, we just have to use the same network config for blob matching and different kernel names for different types. It is possible, but it requires some changes in the host calling code.
This particular PR is not relate to that sort of optimization or improvement and it is focused only on OCL->HIP transition with minimal changes and we can't fit aaaall the improvements into the one PR. I would suggest to create follow-up ticket for that task, because it requires some research and better understanding of the library usecases - it's not obvious what would be better for the user - longer compilation and bigger blob size for all the types or faster compilation and smaller blobs for each actually used type.

I have no ideas what could be the difference, can we ever notice it and all that stuff, so lets limit the changes by the kernel only.

Reply:
> If we want to use one blob for a multiple types, we just have to use the same network config for blob matching

@CAHEK7 No. Blob is identified by filename and build options. The change of network config is not required.

Reply:
> Might be useful to replace these defines with kernels of different types.

@amberhassaan I suspect this does not worth it and recommend evaluating the effect first.